File: src/EditorFeatures/VisualBasicTest/AddAnonymousTypeMemberName/AddAnonymousTypeMemberNameTests.vb
' Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.

Imports Microsoft.CodeAnalysis.CodeFixes
Imports Microsoft.CodeAnalysis.Diagnostics
Imports Microsoft.CodeAnalysis.Editor.VisualBasic.UnitTests.Diagnostics
Imports Microsoft.CodeAnalysis.VisualBasic.AddAnonymousTypeMemberName

Namespace Microsoft.CodeAnalysis.Editor.VisualBasic.UnitTests.AddAnonymousTypeMemberName
    Public Class AddAnonymousTypeMemberNameTests
        Inherits AbstractVisualBasicDiagnosticProviderBasedUserDiagnosticTest

        Friend Overrides Function CreateDiagnosticProviderAndFixer(Workspace As Workspace) As (DiagnosticAnalyzer, CodeFixProvider)
            Return (Nothing, New VisualBasicAddAnonymousTypeMemberNameCodeFixProvider())
        End Function

        <Fact, Trait(Traits.Feature, Traits.Features.CodeActionsAddAnonymousTypeMemberName)>
        Public Async Function Test1() As Task
            Await TestInRegularAndScript1Async(
"
class C
    sub M()
        dim v = new with {[||]me.Equals(1)}
    end sub
end class",
"
class C
    sub M()
        dim v = new with {.{|Rename:V|} = me.Equals(1)}
    end sub
end class")
        End Function

        <Fact, Trait(Traits.Feature, Traits.Features.CodeActionsAddAnonymousTypeMemberName)>
        Public Async Function TestExistingName1() As Task
            Await TestInRegularAndScript1Async(
"
class C
    sub M()
        dim v = new with {.V = 1, [||]me.Equals(1)}
    end sub
end class",
"
class C
    sub M()
        dim v = new with {.V = 1, .{|Rename:V1|} = me.Equals(1)}
    end sub
end class")
        End Function

        <Fact, Trait(Traits.Feature, Traits.Features.CodeActionsAddAnonymousTypeMemberName)>
        Public Async Function TestExistingName2() As Task
            Await TestInRegularAndScript1Async(
"
class C
    sub M()
        dim v = new with {.v = 1, [||]me.Equals(1)}
    end sub
end class",
"
class C
    sub M()
        dim v = new with {.v = 1, .{|Rename:V1|} = me.Equals(1)}
    end sub
end class")
        End Function

        <Fact, Trait(Traits.Feature, Traits.Features.CodeActionsAddAnonymousTypeMemberName)>
        Public Async Function TestFixAll1() As Task
            Await TestInRegularAndScript1Async(
"
class C
    sub M()
        dim v = new with {{|FixAllInDocument:|}new with {me.Equals(1), me.ToString() + 1}}
    end sub
end class",
"
class C
    sub M()
        dim v = new with {.P = new with {.V = me.Equals(1), .V1 = me.ToString() + 1}}
    end sub
end class")
        End Function

        <Fact, Trait(Traits.Feature, Traits.Features.CodeActionsAddAnonymousTypeMemberName)>
        Public Async Function TestFixAll2() As Task
            Await TestInRegularAndScript1Async(
"
class C
{
    sub M()
    {
        dim v = new with {new with {{|FixAllInDocument:|}me.Equals(1), me.ToString() + 1}}
    }
end class",
"
class C
{
    sub M()
    {
        dim v = new with {.P = new with {.V = me.Equals(1), .V1 = me.ToString() + 1}}
    }
end class")
        End Function

        <Fact, Trait(Traits.Feature, Traits.Features.CodeActionsAddAnonymousTypeMemberName)>
        Public Async Function TestFixAll3() As Task
            Await TestInRegularAndScript1Async(
"
class C
    sub M()
        dim v = new with {{|FixAllInDocument:|}new with {me.Equals(1), me.Equals(2)}}
    end sub
end class",
"
class C
    sub M()
        dim v = new with {.P = new with {.V = me.Equals(1), .V1 = me.Equals(2)}}
    end sub
end class")
        End Function
    End Class
End Namespace

File: src/EditorFeatures/CSharpTest/AddAnonymousTypeMemberName/AddAnonymousTypeMemberNameTests.cs
// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.

using System.Threading.Tasks;
using Microsoft.CodeAnalysis.CodeFixes;
using Microsoft.CodeAnalysis.CSharp.AddAnonymousTypeMemberName;
using Microsoft.CodeAnalysis.Diagnostics;
using Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.Diagnostics;
using Microsoft.CodeAnalysis.Test.Utilities;
using Xunit;

namespace Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.AddAnonymousTypeMemberName
{
    public class AddAnonymousTypeMemberNameTests : AbstractCSharpDiagnosticProviderBasedUserDiagnosticTest
    {
        internal override (DiagnosticAnalyzer, CodeFixProvider) CreateDiagnosticProviderAndFixer(Workspace workspace)
            => (null, new CSharpAddAnonymousTypeMemberNameCodeFixProvider());

        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsAddAnonymousTypeMemberName)]
        public async Task Test1()
        {
            await TestInRegularAndScript1Async(
@"
class C
{
    void M()
    {
        var v = new { [||]this.GetType() };
    }
}",
@"
class C
{
    void M()
    {
        var v = new { {|Rename:Type|} = this.GetType() };
    }
}");
        }

        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsAddAnonymousTypeMemberName)]
        public async Task TestExistingName()
        {
            await TestInRegularAndScript1Async(
@"
class C
{
    void M()
    {
        var v = new { Type = 1, [||]this.GetType() };
    }
}",
@"
class C
{
    void M()
    {
        var v = new { Type = 1, {|Rename:Type1|} = this.GetType() };
    }
}");
        }

        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsAddAnonymousTypeMemberName)]
        public async Task TestFixAll1()
        {
            await TestInRegularAndScript1Async(
@"
class C
{
    void M()
    {
        var v = new { {|FixAllInDocument:|}new { this.GetType(), this.ToString() } };
    }
}",
@"
class C
{
    void M()
    {
        var v = new { P = new { Type = this.GetType(), V = this.ToString() } };
    }
}");
        }

        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsAddAnonymousTypeMemberName)]
        public async Task TestFixAll2()
        {
            await TestInRegularAndScript1Async(
@"
class C
{
    void M()
    {
        var v = new { new { {|FixAllInDocument:|}this.GetType(), this.ToString() } };
    }
}",
@"
class C
{
    void M()
    {
        var v = new { P = new { Type = this.GetType(), V = this.ToString() } };
    }
}");
        }

        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsAddAnonymousTypeMemberName)]
        public async Task TestFixAll3()
        {
            await TestInRegularAndScript1Async(
@"
class C
{
    void M()
    {
        var v = new { {|FixAllInDocument:|}new { this.GetType(), this.GetType() } };
    }
}",
@"
class C
{
    void M()
    {
        var v = new { P = new { Type = this.GetType(), Type1 = this.GetType() } };
    }
}");
        }
    }
}
